src/utils/check.hpp

```diff
@@ -188,8 +188,74 @@ template<class T> T& _UsePtr(const std::unique_ptr<T> & ptr, int line, const cha
 	return *ptr;
 }
 
+
+// -------------------------------------------------------------------
+
+/// Where to run invariant check / which invariant (pre,post,both,...)
+typedef enum {
+	e_invariant_place_pre, ///< run check only before function main body
+	e_invariant_place_post, ///< run check only after function main body
+	e_invariant_place_both, ///< run all checks
+} t_invariant_place;
+
+/// This is a RAII that calls Precond() and Postcond() for an object. Use it like a lock-guard in your member functions.
+/// @owner rafal
+template<typename TC> class c_ig final {
+	public:
+		c_ig(const TC & thisobj, t_invariant_place place, int fromline, const char* fromfile)
```

Comment:
ok we could check fromfile to be sure. (though it should be used by macro that pastes there __FILE__)